event-svc/src/event/validator/signed.rs

```diff
@@ -0,0 +1,265 @@
+                                            .unwrap_or_else(|err| panic!("should not have stored invalid data in db for init cid={}. err={:#}", d.id(), err));
+                                let init = std::sync::Arc::new(init);
+                                self.track_init(init_cid, init.clone()); // cache in case other events need it
+                                Some(init)
+                            }
+                            None => {
+                                // if it depends on invalid item, it's invalid, else we have to wait for more info
+                                if self.invalid_init.contains(d.id()) {
+                                    return Ok(ValidationStatus::InvalidData(format!("data event refers to stream with invalid init event (cid={})", d.id())));
+                                } else {
+                                    return Ok(ValidationStatus::Pending);
+                                }
+                            }
+                        }
+                    }
+                } else {
+                    // don't need the init event so we don't even try to look it up
+                    // we could, and if we have it we'd get an extra layer of security but for now
+                    // we're going to avoid the IO in this case
+                    None
+                };
+
+                match self.verify_signature(init, signed, opts).await {
+                    Ok(_) => Ok(ValidationStatus::ValidData),
+                    Err(err) => Ok(ValidationStatus::InvalidData(err)),
+                }
+            }
+            unvalidated::Payload::Init(init) => {
+                // yes, we are the init in this case but it still works
+                match self.verify_signature_against_init(init, signed, opts).await {
+                    Ok(_) => Ok(ValidationStatus::ValidInit),
+                    Err(err) => Ok(ValidationStatus::InvalidInit(err)),
+                }
+            }
+        }
+    }
+
+    async fn verify_signature(
+        &mut self,
+        init: Option<Arc<unvalidated::Event<Ipld>>>,
+        signed: &signed::Event<Ipld>,
+        opts: &VerifyJwsOpts,
+    ) -> std::result::Result<(), String> {
+        match init.as_deref() {
+            Some(unvalidated::Event::Time(_)) => {
+                Err("init event pointer cannot reference a time event".to_string())
+            }
+            Some(unvalidated::Event::Signed(s)) => match s.payload() {
+                unvalidated::Payload::Data(_) => {
+                    Err("init event pointer cannot reference a data event".to_string())
+                }
+                unvalidated::Payload::Init(init) => {
+                    self.verify_signature_against_init(init, signed, opts).await
+                }
+            },
+            Some(unvalidated::Event::Unsigned(init)) => {
+                self.verify_signature_against_init(init, signed, opts).await
+            }
+            None => {
```

Comment:
migrations (when require_init is false) we don't check the controller info against the init (currently)

Reply:
I'm starting to feel like migrations should maybe just bypass validation entirely so we can take out any migration-related special casing out of all of the validation code...

Reply:
yea, I'm coming to that conclusion too (I think I wrote something similar above in another comment).. it seems like using the service for migrations is weird when it literally just skips every piece so it can go straight to the database call

Reply:
adjusted to simply skip doing any work when if we're not requiring validation so this path has been removed. we now just covert things to "valid" if the caller says, "I want this to be implied so give me back things so I can move on"